File: bp-agent/src-go/llm/rotation.go
package llm

import "sync"

// Rotator provides round-robin selection of keys.
type Rotator struct {
	mu   sync.Mutex
	keys []string
	next int
}

// NewRotator creates a new Rotator.
func NewRotator(keys []string) *Rotator {
	return &Rotator{keys: keys}
}

// Next returns the next key in rotation.
func (r *Rotator) Next() string {
	r.mu.Lock()
	defer r.mu.Unlock()
	if len(r.keys) == 0 {
		return ""
	}
	key := r.keys[r.next%len(r.keys)]
	r.next++
	return key
}

File: bp-agent/src-rs/llm/rotation.rs
use std::sync::Mutex;

pub struct Rotator {
    keys: Vec<String>,
    next: Mutex<usize>,
}

impl Rotator {
    pub fn new(keys: Vec<String>) -> Self {
        Self {
            keys,
            next: Mutex::new(0),
        }
    }

    pub fn next(&self) -> Option<String> {
        if self.keys.is_empty() {
            return None;
        }
        let mut idx = self.next.lock().ok()?;
        let key = self.keys[*idx % self.keys.len()].clone();
        *idx += 1;
        Some(key)
    }
}
